### PR TITLE
Incremental Keccak API

### PR DESCRIPTION
**Pull request by ph4r05**
- needed for TREZOR integration

### CODE REVIEW

File: src/crypto/keccak.h [hunk 1]
```diff
@@ -15,6 +15,17 @@
 #define ROTL64(x, y) (((x) << (y)) | ((x) >> (64 - (y))))
 #endif
 
+// SHA3 Algorithm context.
+typedef struct KECCAK_CTX
+{
+  // 1600 bits algorithm hashing state
+  uint64_t hash[25];
+  // 1536-bit buffer for leftovers
+  uint64_t message[24];
```
Comment (moneromooo-monero) [inline src/crypto/keccak.h]: That seems too large. Looks like it needs 136 bytes, so 17 uint64_t. Is this wrong ?

File: src/crypto/keccak.c [hunk 2]
```diff
@@ -132,3 +132,77 @@ void keccak1600(const uint8_t *in, size_t inlen, uint8_t *md)
...
+
+
+void keccak_init(KECCAK_CTX * ctx){
+    memset(ctx, 0, sizeof(KECCAK_CTX));
+}
+
+void keccak_update(KECCAK_CTX * ctx, const uint8_t *in, size_t inlen){
+    size_t idx = (size_t)ctx->rest;
+
+    if (ctx->rest & KECCAK_FINALIZED) return; // too late for additional input
```
Comment (moneromooo-monero) [inline src/crypto/keccak.c]: That might be worth a hard error, it's a programming error.

File: src/crypto/keccak.c [hunk 3]
```diff
@@ -132,3 +132,77 @@ void keccak1600(const uint8_t *in, size_t inlen, uint8_t *md)
...
+
+        in  += left;
+        inlen -= left;
+    }
+    while (inlen >= KECCAK_BLOCKLEN) {
+        uint64_t* aligned_message_block;
+        if (IS_ALIGNED_64(in)) {
+            // the most common case is processing of an already aligned message
+            // without copying it
+            aligned_message_block = (uint64_t*)(void*)in;
```
Comment (moneromooo-monero) [inline src/crypto/keccak.c]: The double cast seems unneeded.

File: src/crypto/keccak.c [hunk 4]
```diff
@@ -132,3 +132,77 @@ void keccak1600(const uint8_t *in, size_t inlen, uint8_t *md)
...
+        if (inlen < left) return;
+
+        // process partial block
+        KECCAK_PROCESS_BLOCK(ctx->hash, ctx->message);
+
+        in  += left;
+        inlen -= left;
+    }
+    while (inlen >= KECCAK_BLOCKLEN) {
+        uint64_t* aligned_message_block;
```
Comment (moneromooo-monero) [inline src/crypto/keccak.c]: Could be const, it's input.

**Comment by ph4r05** [comment]
Thanks for review. I will squash eventually.

File: src/crypto/keccak.c [hunk 5]
```diff
@@ -132,3 +132,77 @@ void keccak1600(const uint8_t *in, size_t inlen, uint8_t *md)
...
+        KECCAK_PROCESS_BLOCK(ctx->hash, ctx->message);
+        ctx->rest = KECCAK_FINALIZED; // mark context as finalized
+    }
+
+    static_assert(KECCAK_BLOCKLEN > KECCAK_DIGESTSIZE, "");
+    if (md) {
+        memcpy(md, ctx->hash, KECCAK_DIGESTSIZE);
+    }
+
+    memset(ctx, 0, sizeof(KECCAK_CTX));
```
[the remark this answers is not in the document]
Comment (ph4r05) [reply src/crypto/keccak.c]: I would leave the call whether necessary or not to you, you know better

File: src/crypto/keccak.c [hunk 6]
```diff
@@ -132,3 +132,77 @@ void keccak1600(const uint8_t *in, size_t inlen, uint8_t *md)
...
+
+        // process partial block
+        KECCAK_PROCESS_BLOCK(ctx->hash, ctx->message);
+
+        in  += left;
+        inlen -= left;
+    }
+    while (inlen >= KECCAK_BLOCKLEN) {
+        uint64_t* aligned_message_block;
+        if (IS_ALIGNED_64(in)) {
```
[the remark this answers is not in the document]
Comment (ph4r05) [reply src/crypto/keccak.c]: Fixed the uint64_t xor and computation of `IS_ALIGNED_64` only once. 

But I don't see how to make unaligned input data aligned as in this state the leftover message buffer is empty and one Keccak round needs 17*8=136 bytes at a time.

File: src/crypto/keccak.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (moneromooo-monero) [reply src/crypto/keccak.c]: Probably unneeded, since there's no free there, though the caller might. We can add it later if it's found to be needed.

**Comment by moneromooo-monero** [comment]
FYI: https://github.com/monero-project/monero/pull/4275

**Comment by ph4r05** [comment]
Aah thanks for the tests! I wanted to add tests after merge.

So next time.

pá 17. 8. 2018 v 18:51 odesílatel moneromooo-monero <
notifications@github.com> napsal:

**Pull request merged**